### ISSUE TITLE
Mixing Dialog and Editor. FocusRing?

### ISSUE THREAD

**Issue by angerman** [jtdaugherty/brick#149]
So I've tried to build the following:
```
.----Add Item---.
| Key: [      ] |
| Val: [      ] |
| [Cancel] [OK] |
'---------------'
```
as a dialog.

Now, while I can use the focus ring between key and value, I fail to see how to compose the focus ring of the inner widget with the dialog, such that the dialog gets focus after the widgets focus ring is exhausted and the widget gets the focus after the dialogs focus ring is exhausted.

That would make the dialog composeable over the widget. Right now I fear I have to ignore the Dialog and just build the full widget by hand?

**Comment by angerman** [comment]
The final solution I came up with was the following:
```
--
-- Build an input widget.
--
-- Name: [ ... ]
-- Key:  [ ... ]
--
--

data Name = Name | Key deriving (Ord, Eq, Show)
data Input = Input
             { _focusRing :: F.FocusRing Name
             , _name :: E.Editor String Name
             , _key  :: E.Editor String Name
             }

makeLenses ''Input

renderInput :: Input -> Widget Name
renderInput i = ui
  where e1 = F.withFocusRing (i ^. focusRing) (E.renderEditor (str . unlines)) (i ^. name)
        e2 = F.withFocusRing (i ^. focusRing) (E.renderEditor (str . unlines)) (i ^. key)

        ui = C.centerLayer $ vLimit 5 $
          (str "Name: " <+> (vLimit 1 $ hLimit 20 e1))
          <=>
          (str "Key:  " <+> (vLimit 1 $ hLimit 20 e2))

handleInputEvent :: V.Event -> Input -> EventM Name (Either Input (Maybe (String, String)))
handleInputEvent (V.EvKey (V.KChar '\t') []) i = return $ Left (i & focusRing %~ F.focusNext)
handleInputEvent (V.EvKey V.KBackTab     []) i = return $ Left (i & focusRing %~ F.focusPrev)
handleInputEvent (V.EvKey V.KEsc         []) i = return $ Right Nothing
handleInputEvent (V.EvKey V.KEnter       []) i = return $ Right $ Just ( filter (/= '\n') . unlines . E.getEditContents $ i ^. name
                                                                       , filter (/= '\n') . unlines . E.getEditContents $ i ^. key)
handleInputEvent ev i = Left <$> case F.focusGetCurrent ( i ^. focusRing) of
  Just Name -> handleEventLensed i name E.handleEditorEvent ev
  Just Key  -> handleEventLensed i key  E.handleEditorEvent ev
  Nothing   -> return i
handleInputEvent _ i = return $ Left i

initialInputState :: Input
initialInputState = Input (F.focusRing [Name, Key])
                          (E.editor Name Nothing "")
                          (E.editor Key  Nothing "")
```

I'm a bit unhappy about the `Either Input (Maybe (String, String))`, as that doesn't really allow to use `handleEventLensed` down the line. On the other hand, if I do not turn the contents into the next state, the component ends up not being self contained. 

But maybe I'm looking at this wrong. The Input Form should have some signature approximating: `getData :: m (Maybe (String, String))`. E.g. either I'm getting data (confirm event) or I'm not getting any input data (cancel event).

If I'm going at this wrong, I'd really appreciate some critique of this approach.

**Comment by jtdaugherty** [comment]
The `Dialog` widget is actually intended for simpler use cases where you just want to ask the user a question and let them pick from a set of responses. But I didn't make that clear in the documentation. Right now there isn't a way to cleanly "chain" focus rings together the way you describe.

At present, I think the cleanest approach here would be to avoid `Dialog` entirely and make your own version of it that includes the right inputs (such as buttons) in your focus ring. It won't take much code at all to reproduce the visual appearance of a `Dialog` but the result will be less complex than trying to re-use a `Dialog` with a set of inputs like you're doing here.

Meanwhile I'm happy to spend some time thinking about ways to make `Dialog` more re-usable. I hope that helps!

**Comment by jtdaugherty** [comment]
@angerman - have you had any luck with an alternative approach? Right now I'd like to keep `Dialog` just focused on its intended purpose -- to give the user a chance to answer a question with a set of responses. That's how it's documented at present. I'm happy to help and take a look at your current approach if that would be useful. Thanks!

**Comment by angerman** [comment]
@jtdaugherty - I have not yet. I have been playing with the new Forms API.  I think the Dialog should just have a stark warning about it’s limitations and that any use outside of the trivial one is better handled using a fully customized widget instead of trying to reuse the dialog.

**Comment by jtdaugherty** [comment]
Okay, I will do that - my hope is that the built-in widgets are useful for some simple purposes, but if not, I also hope they're helpful demonstrations in cases where more sophisticated ones are needed. And my hope is that those more sophisticated ones are relatively easy to build. I'll push out a documentation fix in the next release.

Thanks!
